### PR TITLE
Create interfaces for block marshalling

### PR DESCRIPTION
**Pull request by mtlynch**
Create interface for marshalling type.Block and unmarshalling
consensus.processedBlock. The explicit interfaces will facilitate
testing of the consensus package (specifically the validHeader
function).

### CODE REVIEW

**Comment by mtlynch** [comment]
Next steps, I'd like to replace the `blockMarshaller`, `processedBlockUnmarshaller`, and `dosBlocks` members of `ConsensusSet` with a `blockValidator` interface like:

```
type blockValidator interface {
  validate(types.Block) error
}

type stdBlockValidator struct {
  dosBlocks map[types.BlockID]struct{}
  blockMarshaller blockMarshaller
  processedBlockUnmarshaller processedBlockUnmarshaller
}

func (v stdBlockValidator) validate { 
  // move over current contents of validHeader()
}
```

**Comment by DavidVorick** [comment]
I'm not sure that a BlockUnmarshaller makes sense, using this strategy would lead to creating an typeUnmarshaller type for every single type that gets unmarshalled. This seems like the wrong approach to mocking the encoding package, as we have dozens of types that get encoded throughout Sia.

hmm.

**Comment by DavidVorick** [comment]
It would probably be sufficient to make a generic mock of the encoding package, having a type that has a `Marshal` and `Unmarshal` function implemented. For the stub case, all objects would marshal to []byte{'1'} and all calls to unmarshal would just return whatever object was already being pointed at (maybe panicking or returning an error if 'nil' was passed).

**Comment by mtlynch** [comment]
> It would probably be sufficient to make a generic mock of the encoding package, having a type that has a Marshal and Unmarshal function implemented. For the stub case, all objects would marshal to []byte{'1'} and all calls to unmarshal would just return whatever object was already being pointed at (maybe panicking or returning an error if 'nil' was passed).

That makes sense, pushing a commit now that does this.

File: modules/consensus/block_marshal.go [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+	stdProcessedBlockUnmarshaller struct{}
+)
+
+func (m stdBlockMarshaller) Marshal(block types.Block) []byte {
+	return encoding.Marshal(block)
+}
+
+func (um stdProcessedBlockUnmarshaller) Unmarshal(blockBytes []byte, block *processedBlock) error {
+	return encoding.Unmarshal(blockBytes, block)
+}
```
Comment (DavidVorick) [inline modules/consensus/block_marshal.go]: is this file still needed?

File: modules/consensus/consensusset.go [hunk 2]
```diff
@@ -54,6 +55,9 @@ type ConsensusSet struct {
 	// block.
 	checkingConsistency bool
 
+	// marshaler encodes and decodes between objects and byte slices.
+	marshaler encoding.GenericMarshaler
```
Comment (DavidVorick) [inline modules/consensus/consensusset.go]: interesting approach. Probably a good way to do things

File: modules/consensus/accept.go [hunk 3]
```diff
@@ -61,7 +60,7 @@ func (cs *ConsensusSet) validHeader(tx *bolt.Tx, b types.Block) error {
 	}
 
 	// Check that the block is below the size limit.
-	if uint64(len(encoding.Marshal(b))) > types.BlockSizeLimit {
+	if uint64(len(cs.marshaler.Marshal(b))) > types.BlockSizeLimit {
```
Comment (DavidVorick) [inline modules/consensus/accept.go]: This seems reasonable.

**Comment by DavidVorick** [comment]
Looks good except for the block marshaller file that's still around

File: modules/consensus/block_marshal.go (hunk 1 above)
Comment (mtlynch) [inline modules/consensus/block_marshal.go]: Ah, good catch. Fixed.

**Comment by mtlynch** [comment]
Thanks for reviewing. Fixed the issues you mentioned and rebased onto master.

File: encoding/marshal.go [hunk 4]
```diff
@@ -21,20 +21,34 @@ var (
...
+
+	// StdGenericMarshaler is an implementation of GenericMarshaler that uses
+	// the encoding.Marshal and encoding.Unmarshal functions to perform
+	// its marshaling/unmarshaling.
+	StdGenericMarshaler struct{}
+
+	// An Encoder writes objects to an output stream.
+	Encoder struct {
+		w io.Writer
+	}
```
Comment (lukechampine) [inline encoding/marshal.go]: is this being used anywhere?
